Sharpmake/Builder.cs

```diff
@@ -322,12 +319,15 @@ private void BuildProjectAndSolutionTask(object parameter)
                 throw new Error("error, class type not supported: {0}", type.FullName);
             }
 
-            lock (_buildScheduledType)
+            lock (Arguments.TypesToGenerate)
             {
                 foreach (Type projectDependenciesType in projectDependenciesTypes)
                 {
-                    if (_buildScheduledType.Add(projectDependenciesType))
+                    if (!Arguments.TypesToGenerate.Contains(projectDependenciesType))
```

Comment:
I'd keep _buildScheduledType for the test since it's a hashset, but you can add to TypesToGenerate